cardano-cli/src/Cardano/CLI/Shelley/Commands.hs

```diff
@@ -477,6 +487,24 @@ renderGenesisCmd cmd =
 -- Shelley CLI flag/option data types
 --
 
+data NetworkIdArg
+  = NetworkIdFromCLI NetworkId
+  | NetworkIdFromEnv
+  deriving (Show,Eq)
+
+-- Commands with this argument
+-- behave different depending on whether a network-id option is given or not.
+--
+data SignForNetworkId
+  = SignForNetworkId !NetworkId
+  | TakeNetworkIdFromKey
+  deriving (Show,Eq)
+
+-- Commands with this argument default to mainnet, when no network Id is defined.
+
+newtype NetworkIdOrMainnet = NetworkIdOrMainnet {unNetworkIdOrMainnet :: NetworkIdArg} -- ^ If the lookup of NetworkId fails default to Mainnet
```

Comment:
No I disagree here. We shouldn't just default to mainnet the user must indicate the `NetworkMagic` they are using or we fail loudly.

Reply:
I agree. This should be done in a separate PR.
This PR adds new functionality without breaking the existing API.
This `newtype NetworkIdOrMainnet`  just visibly marks the behavior of the existing code.
It makes the commands  that default to `mainnet` visible, in the existing code !
It does not add new default cases !  It only points at the default cases that already exist.

Reply:
You have added code [here](https://github.com/input-output-hk/cardano-node/pull/4666/files#diff-cc9b567f3acd813c81e105c074ebd783719c74364c88a78c3c22638322354f6dR1171) that defaults to mainnet and I'm saying we should not do this. Therefore we do not need this additional `newtype`.

Reply:
This is the same default case that is already in master [here](https://github.com/input-output-hk/cardano-node/blob/022d552ce8f72f1a2cb19410bbb5243a4ec539a5/cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs#L1183)  just a few lines below.
The old semantic was that when no network id was given in the CLI the function would default to mainnet.
The new semantic is that both the environment variable and a CLI option can overwrite that default.
I tried to add support for the EnvVar without breaking the existing API.
We can still remove that default in a new PR.